server/src/main/java/org/opensearch/common/cache/store/OpenSearchOnHeapCache.java

```diff
@@ -174,18 +178,37 @@ public <K, V> ICache<K, V> create(CacheConfig<K, V> config, CacheType cacheType,
             boolean statsTrackingEnabled = statsTrackingEnabled(config.getSettings(), config.getStatsTrackingEnabled());
             ICacheBuilder<K, V> builder = new Builder<K, V>().setDimensionNames(config.getDimensionNames())
                 .setStatsTrackingEnabled(statsTrackingEnabled)
-                .setMaximumWeightInBytes(((ByteSizeValue) settingList.get(MAXIMUM_SIZE_IN_BYTES_KEY).get(settings)).getBytes())
                 .setExpireAfterAccess(((TimeValue) settingList.get(EXPIRE_AFTER_ACCESS_KEY).get(settings)))
                 .setWeigher(config.getWeigher())
                 .setRemovalListener(config.getRemovalListener());
             Setting<String> cacheSettingForCacheType = CacheSettings.CACHE_TYPE_STORE_NAME.getConcreteSettingForNamespace(
                 cacheType.getSettingPrefix()
             );
+            long maxSizeInBytes = ((ByteSizeValue) settingList.get(MAXIMUM_SIZE_IN_BYTES_KEY).get(settings)).getBytes();
```

Comment:
I am assuming this Cache is initialized during node startup. If not, the value of `MAXIMUM_SIZE_IN_BYTES_KEY` might change during initialization. Also, I am wondering why the `NUMBER_OF_SEGMENTS` is passed via config, and `MAXIMUM_SIZE_IN_BYTES_KEY` is directly read through setting.

Reply:
Just saw there is `getNumberOfSegments()` method with ICacheBuilder. Should we not be using that to add it in CacheConfig?

Reply:
Yeah this might seem a bit confusing, let me try explaining this if I understood your question correctly. 
>Also, I am wondering why the NUMBER_OF_SEGMENTS is passed via config, and MAXIMUM_SIZE_IN_BYTES_KEY is directly read through setting.

CacheConfig is a way to pass certain parameters from core to plugin and vice versa. Each cache maintains its own max size setting(`MAXIMUM_SIZE_IN_BYTES_KEY`) which can configured by the user. So onHeap has its own, disk cache has its own, where tieredCache doesn't as it is just a composition of individual caches.

In case when a user wants to use tiered cache with X segments, tiered cache will need to create X heap/disk cache objects and need to pass this segmentCount to each respective cache implementation(so that each cache allocates each segment `maxSize/X`) which is achieved via CacheConfig. TieredCache isn't aware of each cache's `maxSize` value beforehand as it is managed by each cache's settings, so this one is fetched from setting.

Reply:
>Just saw there is getNumberOfSegments() method with ICacheBuilder. Should we not be using that to add it in CacheConfig?

As mentioned above, CacheConfig is used here to pass numberOfSegments from tieredCache to respective cache implementation either sitting core(like this one) or plugin(like ehcache diskCache). And then cache reads it(via factory) and forms its own `ICacheBuilder` to create a respective object. So we can't use ICacheBuilder to communicate this segmentNumber/count.

Reply:
This is also related to whether we should use maxSize at tieredCache level as well. Will wait on that for us to decide.

Reply:
Seems like this is related to https://github.com/opensearch-project/OpenSearch/pull/16047#discussion_r1785498485. Resolving this one for now, reopen if any issues.